### ISSUE TITLE
bugprone-misplaced-widening-cast does not complain in ArraySubscriptExpr, pointer offsetting

### ISSUE THREAD

**Issue by LebedevRI** [llvm/llvm-project#43234]
|  |  |
| --- | --- |
| Bugzilla Link | [43889](https://llvm.org/bz43889) |
| Version | unspecified |
| OS | Linux |
| CC | @AaronBallman,@JonasToth,@gribozavr,@Xazax-hun |

## Extended Description 
Aha, undoing https://reviews.llvm.org/D32164 (not a fan, and it forgot to
update docs!), helps with 2 out of 5: https://godbolt.org/z/u16A13

**Comment by LebedevRI** [comment]
Was looking at LGTM report (https://lgtm.com/), and to my surprise it complained
about some "Multiplication result converted to larger type", even though
clang/clang-tidy does not complain.

The following patterns are not warned upon by said clang-tidy check:

```c++
bool t0(long a, int b, int c) {
    return a != b * c; // lhs of comparison is `long`, rhs is an int-wide mul
}

int* t1(int* base, long a, int b, int c) {
    return &base[a + b * c]; // int-wide mul is being added to long
}

void sink(long);
void t2(int a, int b) {
    return sink(a * b); // sink() takes long, but we pass int-wide mul
}
```

These two aren't being diagnosed, but i'm wondering whether they should be,
because the result of multiplication will be widened to the pointer size
before being applied as offset.

```c++
int* QUESTION(int* base, int a, int b) {
    return &base[a * b]; // this isn't diagnose, but should it be?
}

int* QUESTION2(int* base, int a, int b) {
    return (base + a * b); // this isn't diagnose, but should it be?
}
```

**Comment by vbvictor** [comment]
https://godbolt.org/z/KrhrWxdsv
